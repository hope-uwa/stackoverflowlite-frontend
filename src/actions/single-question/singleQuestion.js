/* eslint-disable max-len */
import Axios from 'axios';
import {
  SINGLE_QUESTION_FAILURE,
  SINGLE_QUESTION_LOADING,
  SINGLE_QUESTION_SUCCESS,
} from '../types/allquestions';

export const questionLoading = () => ({
  type: SINGLE_QUESTION_LOADING
});
export const questionSuccess = (question, answers, preferred) => ({
  type: SINGLE_QUESTION_SUCCESS,
  question,
  answers,
  preferred
});
export const questionFailure = error => ({
  type: SINGLE_QUESTION_FAILURE,
  error
});
export const loadSingleQuestion = id => (dispatch) => {
  const request = Axios({
    method: 'GET',
    url: `https://uwaelpis.herokuapp.com/api/v1/questions/${id}`,
    headers: [],
  });
  dispatch(questionLoading());
  return request.then(
    (response) => {
      const answerArray = response.data.question.answers;
      const preferred = Array.isArray(answerArray) ? answerArray.filter(x => x.preferred === 'true') : '';
      const answers = Array.isArray(answerArray) ? answerArray.filter(x => x.preferred !== 'true') : '';
      const preferredArray = preferred || [];
      dispatch(questionSuccess(response.data.question, answers, preferredArray));
    },
    err => dispatch(questionFailure(err)),
  );
};
