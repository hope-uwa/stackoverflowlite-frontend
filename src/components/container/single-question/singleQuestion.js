import React from 'react';
import {connect} from 'react-redux';
import {
loadSingleQuestion
} from '../../../actions/single-question/singleQuestion';


export class SingleQuestion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {

    };
  }

  componentDidMount =()=>{
      const {id} = this.props
      this.props.getSingleQuestion(id)
  }
  render(){
      console.log(this.props)
      const { question } = this.props;
      
      return(
           
        <div className="row align-left question-board pl2 pb1 mb1 pls1">
            <div className="col9">
                <div className="question-head mb1">
                    <h1> {question.title}</h1>
                    <span className="ml2">5 days ago by</span><span>Elpis</span>
                </div>
            </div>
            <div className="col3 pt1 pb1">
                <button id="" className="btn btn-success btn-md btn-shadow">
                        Add an Answer
                </button>
            </div>
            <div className="question-body">
                {question.body}
            </div>   
        </div>

      )
  }
}

const mapStateToProps = (state, myOwnProps )=>{
    console.log(state)
    return{
        id: myOwnProps.match.params.id,
        question: state.singleQuestion.question

    }
}
const mapDispatchToProps ={
    getSingleQuestion : (x)=> loadSingleQuestion(x)
}
export default connect(mapStateToProps, mapDispatchToProps)(SingleQuestion);