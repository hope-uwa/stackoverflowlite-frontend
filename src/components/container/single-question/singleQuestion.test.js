import React from 'react';

import { shallow } from 'enzyme';
import { SingleQuestion, mapDispatchToProps, mapStateToProps } from './singleQuestion';
// import { loadAnArticle } from '../../../actions/singlearticle/singleArticleAction';

describe('landing page', () => {
  const getProps = value => (
    {
      id: '',
      question: [],
      answers: [{ answer_body: '', user_name: '' }],
      preferred: [],
      postedAnswer: [{ answer_body: value }],
      getSingleQuestion: () => {},
      chooseAnswer: () => {},
      postAnswer: () => {},
      deletePost: () => {},
    }
  );
  it('should show page is loading', () => {
    const props = getProps('true');
    const component = shallow(<SingleQuestion {...props}/>);
    expect(component).toMatchSnapshot();
  });

  it('should define makePreferred', () => {
    const props = getProps('true');
    const component = shallow(<SingleQuestion {...props}/>);
    component.instance().makePreferred('3', '2');
    expect(component.instance().makePreferred).toBeDefined();
  });

  it('should define deleteAction', () => {
    const props = getProps('true');
    const component = shallow(<SingleQuestion {...props}/>);
    component.instance().deleteAction();
    expect(component.instance().deleteAction).toBeDefined();
  });

  it('should load componentWillRecieveProps', () => {
    const props2 = getProps('false');
    const props = getProps('true');
    const component = shallow(<SingleQuestion {...props}/>);
    component.instance().componentWillReceiveProps({ ...props2 });
    expect(component.instance().componentWillReceiveProps).toBeDefined();
  });

  it('define mapStateToProps', () => {
    const state = { singleQuestion: { questions: [] }, postAnswerReducer: { answer: [] } };
    const myOwnProps = { match: { params: { id: '' } } };
    expect(mapStateToProps(state, myOwnProps)).toBeDefined();
  });
  it('define mapDispatchToProps', () => {
    expect(mapDispatchToProps).toBeDefined();
    expect(mapDispatchToProps.getSingleQuestion()).toBeDefined();
    expect(mapDispatchToProps.chooseAnswer()).toBeDefined();
    expect(mapDispatchToProps.postAnswer()).toBeDefined();
    expect(mapDispatchToProps.deletePost()).toBeDefined();
  });
});
