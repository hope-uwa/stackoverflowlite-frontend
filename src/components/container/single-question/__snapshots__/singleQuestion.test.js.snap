// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`landing page should show page is loading 1`] = `
<div
  className="row align-left question-board pl2 pb1 mb1 pls1"
>
  <div
    className="col9"
  >
    <div
      className="question-head mb1"
    >
      <h1>
         
        this is the title
      </h1>
      <span
        className="ml2"
      >
        5 days ago by
      </span>
      <span>
        Elpis
      </span>
    </div>
  </div>
  <div
    className="col3 pt1 pb1"
  >
    <button
      className="btn btn-success btn-md btn-shadow"
      id=""
    >
      Add an Answer
    </button>
  </div>
  <div
    className="question-body"
  />
</div>
`;
